### PR TITLE
mds: display scrub status in ceph status

### PR DESCRIPTION
**Pull request by vshankar**
This introduces changes that allow normal ceph services (osd, mds, etc..) to register with ceph manager via `MgrClient::service_daemon_register()`. Also, services can forward the status of executing tasks to ceph manager. The `task status` makes its way upto the `ServiceMap`.

@batrick @liewegas please take a look.

Sample vstart display output when running mds scrub in a vstart cluster.
```
$ ceph -s
  cluster:                                                          
    id:     18a4f74f-7a1b-45bb-b971-4399c32797be                                                          
    health: HEALTH_ERR                                                                                    
            1 MDSs report damaged metadata                                                                 
                                                                                                           
  services:                                                                                               
    mon: 1 daemons, quorum a (age 3m)    
    mgr: x(active, since 3m)                    
    mds: a:1 {0=a=up:active} 2 up:standby
    osd: 3 osds: 3 up (since 92s), 3 in (since 12h)                 
                                                                    
  task status:                                                                                            
    scrub status:                                                                                         
        mds.0: active [paths:/]                     
                                                                    
  data:                                                                                                   
    pools:   2 pools, 24 pgs                                                                              
    objects: 66.79k objects, 1.4 GiB
    usage:   12 GiB used, 21 GiB / 33 GiB avail             
    pgs:     24 active+clean                                        
                                                                                                          
  io:                                                                                                     
    client:   1.5 MiB/s rd, 1.24k op/s rd, 0 op/s wr
```

### CODE REVIEW

File: src/mds/MDSRank.cc [hunk 1]
```diff
@@ -3595,3 +3608,35 @@ epoch_t MDSRank::get_osd_epoch() const
   return objecter->with_osdmap(std::mem_fn(&OSDMap::get_epoch));  
 }
 
+void MDSRank::get_task_status(std::map<std::string, std::string> *status) {
+  dout(20) << __func__ << dendl;
+
+  // scrub summary for now..
+  std::string scrub_summary = scrubstack->scrub_summary();
+  status->emplace(SCRUB_STATUS_KEY, scrub_summary);
```
Comment (batrick) [inline src/mds/MDSRank.cc]: nit: `std::move(scrub_summary)`

File: src/mds/MDSRank.h [hunk 2]
```diff
@@ -577,6 +580,13 @@ class MDSRank {
 private:
     mono_time starttime = mono_clock::zero();
 
+    // "task" string that gets displayed in ceph status
+    const std::string SCRUB_STATUS_KEY = "scrub status";
```
Comment (batrick) [inline src/mds/MDSRank.h]: `inline static const std::string SCRUB_STATUS_KEY = "scrub status";` would be better.

Comment (batrick) [reply src/mds/MDSRank.h]: See also: https://en.cppreference.com/w/cpp/language/static

File: src/mds/ScrubStack.cc [hunk 3]
```diff
@@ -528,6 +528,52 @@ bool ScrubStack::scrub_in_transition_state() {
   return false;
 }
 
+std::string ScrubStack::scrub_summary() {
+  ceph_assert(mdcache->mds->mds_lock.is_locked_by_me());
+
+  bool have_more = false;
+  std::stringstream ss;
```
Comment (batrick) [inline src/mds/ScrubStack.cc]: These are rather expensive to construct. Let's instead use `CachedStackStringStream`.

File: src/mgr/ServiceMap.h [hunk 4]
```diff
@@ -64,6 +67,32 @@ struct ServiceMap {
       return ss.str();
     }
 
+    std::string get_task_summary(const std::string &task_prefix) const {
```
Comment (batrick) [inline src/mgr/ServiceMap.h]: string_view arg would be better

File: src/mgr/ServiceMap.h [hunk 5]
```diff
@@ -64,6 +67,32 @@ struct ServiceMap {
...
+      // contruct a map similar to:
+      //     {"service1 status" -> {"service1.0" -> "running"}}
+      //     {"service2 status" -> {"service2.0" -> "idle"},
+      //                           {"service2.1" -> "running"}}
+      std::map<std::string, std::map<std::string, std::string>> by_task;
+      for (auto p : daemons) {
```
Comment (batrick) [inline src/mgr/ServiceMap.h]: `const auto& p`

File: src/mgr/ServiceMap.h [hunk 6]
```diff
@@ -64,6 +67,32 @@ struct ServiceMap {
...
+        std::string d = task_prefix + "." + p.first;
+        for (auto q : p.second.task_status) {
```
Comment (batrick) [inline src/mgr/ServiceMap.h]: ditto

File: src/mgr/ServiceMap.h [hunk 7]
```diff
@@ -64,6 +67,32 @@ struct ServiceMap {
...
+          auto p1 = by_task.emplace(q.first, std::map<std::string, std::string>{}).first;
+          auto p2 = p1->second.emplace(d, std::string()).first;
+          p2->second = q.second;
+        }
+      }
+
+      std::stringstream ss;
+      for (auto p : by_task) {
```
Comment (batrick) [inline src/mgr/ServiceMap.h]: ditto

File: src/mgr/ServiceMap.h [hunk 8]
```diff
@@ -64,6 +67,32 @@ struct ServiceMap {
...
+        ss << "\n    " << p.first << ":";
+        for (auto q : p.second) {
```
Comment (batrick) [inline src/mgr/ServiceMap.h]: ditto

File: src/mds/ScrubStack.cc (hunk 3 above)
Comment (vshankar) [reply src/mds/ScrubStack.cc]: ack

File: src/mds/MDSRank.h (hunk 2 above)
Comment (vshankar) [reply src/mds/MDSRank.h]: ack

**Comment by vshankar** [comment]
> Are we also logging to the cluster log? That would also be appropriate.

scrub status in the mds?

> 
> Otherwise looks good. A test would be nice but it may be too racy for small datasets?

Extended scrub tests to cover scrub task summary via ceph status.

**Comment by vshankar** [comment]
jenkins render docs

**Comment by batrick** [comment]
> > Are we also logging to the cluster log? That would also be appropriate.
> 
> scrub status in the mds?

Maybe just new scrubs and major events (paused, finished). I think you can figure that out from the difference between task_status updates?

> > Otherwise looks good. A test would be nice but it may be too racy for small datasets?
> 
> Extended scrub tests to cover scrub task summary via ceph status.

Thanks!

Change this to a formal PR when you feel it's ready for QA.

**Comment by vshankar** [comment]
> Maybe just new scrubs and major events (paused, finished). I think you can figure that out from the difference between task_status updates?

for such events, it would be better to log when the appropriate scrub control operations are triggered.

**Comment by batrick** [comment]
retest this please

**Comment by batrick** [comment]
@vshankar please rebase as I think the master you're based on has an rgw build issue

**Comment by vshankar** [comment]
@batrick rebased and updated

**Comment by batrick** [comment]
@vshankar looks like it must be your PR. Can you investigate:

https://jenkins.ceph.com/job/ceph-pull-requests/30209/console

**Comment by vshankar** [comment]
sure

**Comment by vshankar** [comment]
> @vshankar looks like it must be your PR. Can you investigate:
> 
> https://jenkins.ceph.com/job/ceph-pull-requests/30209/console

`DaemonServer::handle_report()` and `DaemonServer::got_service_map()` can run concurrently. `got_service_map()` (which acquires DaemonServer::lock`) can remove an entry from `DaemonState` while `handle_report()` is accessing it (as it runs lockless).

Looks to me that nothing (in teuthology tests) was exercising this flow till now.

**Comment by vshankar** [comment]
@batrick fixed -- please take a look. `handle_report()` now acquires appropriate locks.

**Comment by vshankar** [comment]
retest this please

**Comment by vshankar** [comment]
jenkins render docs

**Comment by batrick** [comment]
retest this please

File: src/messages/MMgrReport.h [hunk 9]
```diff
@@ -73,7 +73,7 @@ WRITE_CLASS_ENCODER(PerfCounterType)
 
 class MMgrReport : public Message {
 private:
-  static constexpr int HEAD_VERSION = 7;
+  static constexpr int HEAD_VERSION = 8;
```
Comment (sebastian-philipp) [inline src/messages/MMgrReport.h]: just fyi, this will conflict with #29214

**Comment by batrick** [comment]
retest this please

File: src/mgr/ServiceMap.h [hunk 10]
```diff
@@ -12,17 +12,20 @@
...
+#include <boost/optional.hpp>
+
 namespace ceph {
   class Formatter;
 }
 
 struct ServiceMap {
   struct Daemon {
-    uint64_t gid = 0;
+    boost::optional<uint64_t> gid;
```
Comment (tchaikov) [inline src/mgr/ServiceMap.h]: this change is not backward compatible. it breaks `readable.sh`.

```
104: **** failed to decode /var/ssd/ceph/ceph-object-corpus/archive/15.0.0-539-g191ab33faf/objects/ServiceMap::Service/08800c908429da57f2b09b5cc8463080 ****
104: error: buffer::malformed_input: void entity_addr_t::decode(ceph::buffer::v14_2_0::list::const_iterator&) no longer understand old encoding version 1 < 28
104: error: buffer::malformed_input: void entity_addr_t::decode(ceph::buffer::v14_2_0::list::const_iterator&) no longer understand old encoding version 1 < 28
104: **** failed to decode /var/ssd/ceph/ceph-object-corpus/archive/15.0.0-539-g191ab33faf/objects/ServiceMap/074825661e8e8a3fbcb1705c109ad5db ****
104: error: buffer::malformed_input: void entity_addr_t::decode(ceph::buffer::v14_2_0::list::const_iterator&) no longer understand old encoding version 1 < 28
104: error: buffer::malformed_input: void entity_addr_t::decode(ceph::buffer::v14_2_0::list::const_iterator&) no longer understand old encoding version 1 < 28
104: error: buffer::malformed_input: void entity_addr_t::decode(ceph::buffer::v14_2_0::list::const_iterator&) no longer understand old encoding version 1 < 28
```

@vshankar @batrick  https://tracker.ceph.com/issues/41424

Comment (batrick) [reply src/mgr/ServiceMap.h]: Thanks for catching that @tchaikov

**Pull request merged**